File: frontend/worker/strategy-quality.ts
export type JsonRecord = Record<string, unknown>;

export interface StrategyIssue {
  code: string;
  path: string;
  message: string;
  severity: "low" | "medium" | "high";
  blocking: boolean;
}

const marketAssertion = /\b(?:unlike|market demand|users? prefer|customers? prefer|willingness to pay|willing to pay|existing (?:\w+\s+){0,2}(?:tools|apps|products)|(?:tools|apps|competitors?)\b.{0,40}\b(?:fail|fails|lack|lacks|cannot|can't|may not|do not|does not|are not|cause|causes))\b/i;
const unsupportedStrength = /\b(?:many|most|often|frequently|significantly|better than|worse than)\b/i;
const hypothesisMarker = /\b(?:may|might|could|hypothesis|to test)\b/i;
const measurableSignal = /(?:\d|%|percent|at least|at most|no more than|fewer than|more than|within \d)/i;

const basisRank: Record<string, number> = { behavioral_hypothesis: 1, contextual_inference: 2, explicit_brief: 3 };
const confidenceRank: Record<string, number> = { low: 1, medium: 2, high: 3 };

export function isRecord(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function clone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function* traceableItems(value: unknown, path: string): Generator<[string, JsonRecord]> {
  if (isRecord(value)) {
    if (Array.isArray(value.source_refs)) yield [path, value];
    for (const [key, child] of Object.entries(value)) yield* traceableItems(child, `${path}.${key}`);
  } else if (Array.isArray(value)) {
    for (const [index, child] of value.entries()) yield* traceableItems(child, `${path}.${index}`);
  }
}

function resolveRef(reference: string, roots: JsonRecord): { valid: boolean; evidence?: JsonRecord } {
  const parts = reference.split(".");
  let current: unknown = roots[parts[0]];
  if (current === undefined) return { valid: false };
  let evidence: JsonRecord | undefined;
  for (const part of parts.slice(1)) {
    if (isRecord(current) && isRecord(current.evidence)) evidence = current.evidence;
    if (isRecord(current) && part in current) current = current[part];
    else if (Array.isArray(current) && /^\d+$/.test(part) && Number(part) < current.length) current = current[Number(part)];
    else return { valid: false };
  }
  if (isRecord(current) && isRecord(current.evidence)) evidence = current.evidence;
  if (parts[0] === "context" && !evidence) {
    const inferred = ["product_category", "assumptions", "ambiguities"].includes(parts[1]);
    evidence = { basis: inferred ? "contextual_inference" : "explicit_brief", confidence: inferred ? "medium" : "high" };
  }
  return { valid: true, evidence };
}

function itemText(item: JsonRecord): string {
  const ignored = new Set(["source_refs", "evidence", "priority", "support_status"]);
  return Object.entries(item).flatMap(([key, value]) => !ignored.has(key) && typeof value === "string" ? [value] : []).join(" ");
}

export function normalizeStrategyClaims(marketValue: unknown, propositionValue: unknown) {
  const market = clone(marketValue);
  const value = clone(propositionValue);
  let revisionCount = 0;
  for (const [root, payload] of [["market_hypothesis", market], ["value_proposition", value]] as const) {
    for (const [, item] of traceableItems(payload, root)) {
      if (!isRecord(item.evidence) || item.evidence.basis === "explicit_brief") continue;
      for (const [key, content] of Object.entries(item)) {
        if (["source_refs", "evidence", "priority", "support_status"].includes(key) || typeof content !== "string") continue;
        if (!hypothesisMarker.test(content) && (marketAssertion.test(content) || unsupportedStrength.test(content))) {
          item[key] = `Hypothesis to test — ${content}`;
          revisionCount += 1;
        }
      }
    }
  }
  return { market, value, revisionCount };
}

export function evaluateStrategyQuality(
  brief: JsonRecord,
  context: unknown,
  userInsight: unknown,
  market: unknown,
  value: unknown,
  autoRevisionCount = 0,
) {
  const contextData = isRecord(context) ? context : {};
  const userData = isRecord(userInsight) ? userInsight : {};
  const marketData = isRecord(market) ? market : {};
  const valueData = isRecord(value) ? value : {};
  const roots: JsonRecord = { context: contextData, user_insight: userData, market_hypothesis: marketData };
  const issues: StrategyIssue[] = [];

  const requiredMarketArrays = ["current_alternatives", "behavior_hypotheses", "main_risks", "validation_priorities"];
  const requiredValueArrays = ["functional_values", "emotional_values", "social_values", "reasons_to_believe", "message_pillars", "objections", "message_tests"];
  const structurePassed = ["opportunity_statement", "growth_wedge", "competitive_frame"].every((key) => isRecord(marketData[key]))
    && requiredMarketArrays.every((key) => Array.isArray(marketData[key]) && (marketData[key] as unknown[]).length > 0)
    && isRecord(valueData.primary_value)
    && requiredValueArrays.every((key) => Array.isArray(valueData[key]) && (valueData[key] as unknown[]).length > 0);
  if (!structurePassed) issues.push({ code: "structure_contract", path: "strategy", message: "A required V0.3 strategy module or collection is missing.", severity: "high", blocking: true });

  const traceable = [
    ...traceableItems(marketData, "market_hypothesis"),
    ...traceableItems(valueData, "value_proposition"),
  ];
  let referenceIssues = 0;
  let evidenceIssues = 0;
  let claimIssues = 0;
  for (const [path, item] of traceable) {
    const allowedRoots = path.startsWith("market_hypothesis") ? new Set(["context", "user_insight"]) : new Set(["context", "user_insight", "market_hypothesis"]);
    const upstreamEvidence: JsonRecord[] = [];
    for (const ref of (item.source_refs as unknown[]).filter((candidate): candidate is string => typeof candidate === "string")) {
      const root = ref.split(".", 1)[0];
      const resolved = resolveRef(ref, roots);
      if (!allowedRoots.has(root) || !resolved.valid) {
        referenceIssues += 1;
        issues.push({ code: "invalid_source_ref", path: `${path}.source_refs`, message: `${ref} does not resolve to an allowed upstream field.`, severity: "high", blocking: true });
      } else if (resolved.evidence) upstreamEvidence.push(resolved.evidence);
    }
    if (isRecord(item.evidence) && upstreamEvidence.length) {
      const downstreamBasis = basisRank[String(item.evidence.basis)] ?? 0;
      const downstreamConfidence = confidenceRank[String(item.evidence.confidence)] ?? 0;
      const weakestBasis = Math.min(...upstreamEvidence.map((evidence) => basisRank[String(evidence.basis)] ?? 0));
      const weakestConfidence = Math.min(...upstreamEvidence.map((evidence) => confidenceRank[String(evidence.confidence)] ?? 0));
      if (downstreamBasis > weakestBasis || downstreamConfidence > weakestConfidence) {
        evidenceIssues += 1;
        issues.push({ code: "evidence_stronger_than_source", path: `${path}.evidence`, message: "Downstream evidence is stronger than at least one cited source.", severity: "high", blocking: true });
      }
    }
    if (isRecord(item.evidence)) {
      const text = itemText(item);
      const risky = marketAssertion.test(text) || unsupportedStrength.test(text);
      if (item.evidence.basis === "explicit_brief" && risky) {
        claimIssues += 1;
        issues.push({ code: "unsupported_explicit_market_claim", path, message: "A factual market or comparative claim is not established by the submitted brief.", severity: "high", blocking: true });
      } else if (risky && !hypothesisMarker.test(text)) {
        claimIssues += 1;
        issues.push({ code: "unsupported_claim_language", path, message: "Use neutral or explicit hypothesis language for this market claim.", severity: "medium", blocking: false });
      }
    }
  }

  let continuityIssues = 0;
  const primaryValue = isRecord(valueData.primary_value) ? valueData.primary_value : {};
  const primaryRefs = Array.isArray(primaryValue.source_refs) ? primaryValue.source_refs.filter((ref): ref is string => typeof ref === "string") : [];
  const primaryRoots = new Set(primaryRefs.map((ref) => ref.split(".", 1)[0]));
  if (!primaryRoots.has("user_insight") || !primaryRoots.has("market_hypothesis")) {
    continuityIssues += 1;
    issues.push({ code: "primary_value_missing_decision_link", path: "value_proposition.primary_value.source_refs", message: "Primary value must cite both a user insight and a market hypothesis.", severity: "high", blocking: true });
  }
  if (contextData.primary_goal !== brief.business_goal) {
    continuityIssues += 1;
    issues.push({ code: "business_goal_drift", path: "context.primary_goal", message: "The normalized business goal differs from the submitted brief.", severity: "high", blocking: true });
  }

  let testabilityIssues = 0;
  const priorities = Array.isArray(marketData.validation_priorities) ? marketData.validation_priorities : [];
  priorities.forEach((priority, index) => {
    if (!isRecord(priority) || typeof priority.pass_signal !== "string" || typeof priority.fail_signal !== "string" || !measurableSignal.test(priority.pass_signal) || !measurableSignal.test(priority.fail_signal)) {
      testabilityIssues += 1;
      issues.push({ code: "validation_signal_not_measurable", path: `market_hypothesis.validation_priorities.${index}`, message: "Pass and fail signals need an observable threshold, count, percentage, or time bound.", severity: "medium", blocking: false });
    }
  });

  const checks = [
    { code: "structure_contract", label: "Structure contract", status: structurePassed ? "passed" : "warning", detail: structurePassed ? "All four strategy objects passed the V0.3 contract." : "A required strategy object needs review." },
    { code: "reference_integrity", label: "Reference integrity", status: referenceIssues ? "warning" : "passed", detail: referenceIssues ? `${referenceIssues} invalid source reference(s) found.` : "Every source reference resolves to an allowed upstream field." },
    { code: "evidence_inheritance", label: "Evidence inheritance", status: evidenceIssues ? "warning" : "passed", detail: evidenceIssues ? `${evidenceIssues} item(s) overstate upstream evidence.` : "Downstream evidence does not exceed its weakest source." },
    { code: "market_claim_grounding", label: "Market claim grounding", status: claimIssues ? "warning" : "passed", detail: claimIssues ? `${claimIssues} market claim(s) need review.` : autoRevisionCount ? `${autoRevisionCount} inferred claim(s) were reframed before this check.` : "No unsupported factual market claims were detected." },
    { code: "decision_continuity", label: "Decision continuity", status: continuityIssues ? "warning" : "passed", detail: continuityIssues ? `${continuityIssues} upstream decision link(s) need review.` : "Goal, user insight, market hypothesis, and primary value remain connected." },
    { code: "validation_testability", label: "Validation testability", status: testabilityIssues ? "warning" : "passed", detail: testabilityIssues ? `${testabilityIssues} validation plan(s) need measurable signals.` : "Every validation priority has measurable pass and fail signals." },
  ] as Array<{ code: string; label: string; status: "passed" | "warning"; detail: string }>;
  const blockingCount = issues.filter((issue) => issue.blocking).length;
  return {
    status: blockingCount ? "review_required" : issues.length ? "passed_with_notes" : "passed",
    issue_count: issues.length,
    blocking_issue_count: blockingCount,
    auto_revision_count: autoRevisionCount,
    checks,
    issues,
  };
}

File: backend/app/strategy_quality.py
from __future__ import annotations

import re
from collections.abc import Iterator
from typing import Any

from .models import (
    GrowthBrief,
    InsightEvidence,
    MarketHypothesis,
    NormalizedContext,
    QualityCheck,
    StrategyQualityIssue,
    StrategyQualityReview,
    UserInsight,
    ValueProposition,
)


MARKET_ASSERTION_PATTERN = re.compile(
    r"\b(?:unlike|market demand|users? prefer|customers? prefer|willingness to pay|"
    r"willing to pay|existing (?:\w+\s+){0,2}(?:tools|apps|products)|"
    r"(?:tools|apps|competitors?)\b.{0,40}\b(?:fail|fails|lack|lacks|cannot|can't|"
    r"may not|do not|does not|are not|cause|causes))\b",
    re.IGNORECASE,
)
UNSUPPORTED_STRENGTH_PATTERN = re.compile(
    r"\b(?:many|most|often|frequently|significantly|better than|worse than)\b",
    re.IGNORECASE,
)
HYPOTHESIS_MARKER = re.compile(r"\b(?:may|might|could|hypothesis|to test)\b", re.IGNORECASE)
MEASURABLE_SIGNAL = re.compile(
    r"(?:\d|%|percent|at least|at most|no more than|fewer than|more than|within \d)",
    re.IGNORECASE,
)
TOKEN_PATTERN = re.compile(r"[a-z0-9]+", re.IGNORECASE)

BASIS_RANK = {
    "behavioral_hypothesis": 1,
    "contextual_inference": 2,
    "explicit_brief": 3,
}
CONFIDENCE_RANK = {"low": 1, "medium": 2, "high": 3}


def _issue(
    code: str,
    path: str,
    message: str,
    *,
    severity: str = "medium",
    blocking: bool = False,
) -> StrategyQualityIssue:
    return StrategyQualityIssue(
        code=code,
        path=path,
        message=message,
        severity=severity,
        blocking=blocking,
    )


def _traceable_items(value: Any, path: str) -> Iterator[tuple[str, dict[str, Any]]]:
    if isinstance(value, dict):
        if isinstance(value.get("source_refs"), list):
            yield path, value
        for key, child in value.items():
            yield from _traceable_items(child, f"{path}.{key}")
    elif isinstance(value, list):
        for index, child in enumerate(value):
            yield from _traceable_items(child, f"{path}.{index}")


def _evidenced_user_items(
    value: Any, path: str = "user_insight"
) -> Iterator[tuple[str, dict[str, Any]]]:
    if isinstance(value, dict):
        if isinstance(value.get("evidence"), dict):
            yield path, value
        for key, child in value.items():
            yield from _evidenced_user_items(child, f"{path}.{key}")
    elif isinstance(value, list):
        for index, child in enumerate(value):
            yield from _evidenced_user_items(child, f"{path}.{index}")


def _resolve_ref(reference: str, roots: dict[str, Any]) -> tuple[bool, Any, InsightEvidence | None]:
    parts = reference.split(".")
    current = roots.get(parts[0])
    if current is None:
        return False, None, None

    inherited_evidence: InsightEvidence | None = None
    for part in parts[1:]:
        if isinstance(current, dict) and isinstance(current.get("evidence"), dict):
            inherited_evidence = InsightEvidence.model_validate(current["evidence"])
        if isinstance(current, dict) and part in current:
            current = current[part]
        elif isinstance(current, list) and part.isdigit() and int(part) < len(current):
            current = current[int(part)]
        else:
            return False, None, None

    if isinstance(current, dict) and isinstance(current.get("evidence"), dict):
        inherited_evidence = InsightEvidence.model_validate(current["evidence"])
    if parts[0] == "context" and inherited_evidence is None:
        inferred_fields = {"product_category", "assumptions", "ambiguities"}
        basis = (
            "contextual_inference"
            if len(parts) > 1 and parts[1] in inferred_fields
            else "explicit_brief"
        )
        inherited_evidence = InsightEvidence(
            basis=basis,
            confidence="medium" if basis == "contextual_inference" else "high",
            validation_status="needs_validation"
            if basis == "contextual_inference"
            else "brief_supported",
        )
    return True, current, inherited_evidence


def _item_text(item: dict[str, Any]) -> str:
    ignored = {"source_refs", "evidence", "priority", "support_status"}
    return " ".join(
        str(value) for key, value in item.items() if key not in ignored and isinstance(value, str)
    )


def _tokens(text: str) -> set[str]:
    stop = {"the", "a", "an", "and", "or", "to", "for", "of", "in", "on", "with", "is", "are"}
    return {token.lower() for token in TOKEN_PATTERN.findall(text) if token.lower() not in stop}


def normalize_strategy_claim_language(
    market_hypothesis: MarketHypothesis,
    value_proposition: ValueProposition,
) -> tuple[MarketHypothesis, ValueProposition, int]:
    """Make inferred market assertions visibly testable before quality review.

    The rewrite is intentionally conservative: it never upgrades evidence or
    changes the substance of a generated claim. Explicit-brief claims are left
    untouched so a wrongly attributed factual claim still blocks publication.
    """
    market_data = market_hypothesis.model_dump(mode="json")
    value_data = value_proposition.model_dump(mode="json")
    revision_count = 0

    for root_name, payload in (
        ("market_hypothesis", market_data),
        ("value_proposition", value_data),
    ):
        for _, item in _traceable_items(payload, root_name):
            raw_evidence = item.get("evidence")
            if not isinstance(raw_evidence, dict):
                continue
            evidence = InsightEvidence.model_validate(raw_evidence)
            if evidence.basis == "explicit_brief":
                continue
            for key, value in item.items():
                if key in {"source_refs", "evidence", "priority", "support_status"}:
                    continue
                if not isinstance(value, str) or HYPOTHESIS_MARKER.search(value):
                    continue
                if MARKET_ASSERTION_PATTERN.search(value) or UNSUPPORTED_STRENGTH_PATTERN.search(
                    value
                ):
                    item[key] = f"Hypothesis to test — {value}"
                    revision_count += 1

    return (
        MarketHypothesis.model_validate(market_data),
        ValueProposition.model_validate(value_data),
        revision_count,
    )


def evaluate_strategy_quality(
    brief: GrowthBrief,
    context: NormalizedContext,
    user_insight: UserInsight,
    market_hypothesis: MarketHypothesis,
    value_proposition: ValueProposition,
    auto_revision_count: int = 0,
) -> StrategyQualityReview:
    """Validate reference integrity, evidence inheritance, and decision continuity."""
    issues: list[StrategyQualityIssue] = []
    context_data = context.model_dump(mode="json")
    user_data = user_insight.model_dump(mode="json")
    market_data = market_hypothesis.model_dump(mode="json")
    value_data = value_proposition.model_dump(mode="json")
    roots = {
        "context": context_data,
        "user_insight": user_data,
        "market_hypothesis": market_data,
    }

    traceable = [
        *_traceable_items(market_data, "market_hypothesis"),
        *_traceable_items(value_data, "value_proposition"),
    ]
    reference_issues = 0
    evidence_issues = 0
    claim_issues = 0
    grounding_issues = 0

    for path, item in traceable:
        allowed_roots = (
            {"context", "user_insight"}
            if path.startswith("market_hypothesis")
            else {"context", "user_insight", "market_hypothesis"}
        )
        source_evidence: list[InsightEvidence] = []
        for reference in item["source_refs"]:
            if reference.split(".", 1)[0] not in allowed_roots:
                reference_issues += 1
                issues.append(
                    _issue(
                        "invalid_source_root",
                        f"{path}.source_refs",
                        f"{reference} is not an allowed upstream source for this module.",
                        severity="high",
                        blocking=True,
                    )
                )
                continue
            valid, _, evidence = _resolve_ref(reference, roots)
            if not valid:
                reference_issues += 1
                issues.append(
                    _issue(
                        "invalid_source_ref",
                        f"{path}.source_refs",
                        f"{reference} does not resolve to an upstream field.",
                        severity="high",
                        blocking=True,
                    )
                )
            elif evidence is not None:
                source_evidence.append(evidence)

        item_evidence = item.get("evidence")
        if item_evidence and source_evidence:
            downstream = InsightEvidence.model_validate(item_evidence)
            weakest_basis = min(BASIS_RANK[source.basis] for source in source_evidence)
            weakest_confidence = min(
                CONFIDENCE_RANK[source.confidence] for source in source_evidence
            )
            if (
                BASIS_RANK[downstream.basis] > weakest_basis
                or CONFIDENCE_RANK[downstream.confidence] > weakest_confidence
            ):
                evidence_issues += 1
                issues.append(
                    _issue(
                        "evidence_stronger_than_source",
                        f"{path}.evidence",
                        "Downstream evidence is stronger than at least one cited upstream source.",
                        severity="high",
                        blocking=True,
                    )
                )

        text = _item_text(item)
        if item_evidence:
            evidence = InsightEvidence.model_validate(item_evidence)
            risky_market_claim = MARKET_ASSERTION_PATTERN.search(text)
            unsupported_strength = UNSUPPORTED_STRENGTH_PATTERN.search(text)
            if evidence.basis == "explicit_brief" and (risky_market_claim or unsupported_strength):
                claim_issues += 1
                issues.append(
                    _issue(
                        "unsupported_explicit_market_claim",
                        path,
                        "This comparative or market claim is labeled as explicit brief evidence but is not directly established by the brief.",
                        severity="high",
                        blocking=True,
                    )
                )
            elif (risky_market_claim or unsupported_strength) and not HYPOTHESIS_MARKER.search(
                text
            ):
                claim_issues += 1
                issues.append(
                    _issue(
                        "unsupported_claim_language",
                        path,
                        "Use neutral or explicit hypothesis language for frequency, comparative, or causal wording.",
                    )
                )

    for path, item in _evidenced_user_items(user_data):
        evidence = InsightEvidence.model_validate(item["evidence"])
        text = _item_text(item)
        if evidence.basis == "explicit_brief" and MARKET_ASSERTION_PATTERN.search(text):
            claim_issues += 1
            issues.append(
                _issue(
                    "unsupported_explicit_market_claim",
                    path,
                    "The user-insight item introduces a market or alternative claim that is not explicit in the brief.",
                    severity="high",
                    blocking=True,
                )
            )

    primary_refs = value_proposition.primary_value.source_refs
    primary_roots = {reference.split(".", 1)[0] for reference in primary_refs}
    if not {"user_insight", "market_hypothesis"}.issubset(primary_roots):
        grounding_issues += 1
        issues.append(
            _issue(
                "primary_value_missing_decision_link",
                "value_proposition.primary_value.source_refs",
                "Primary value must cite both a user insight and a market hypothesis.",
                severity="high",
                blocking=True,
            )
        )

    segment_tokens = _tokens(user_insight.target_user.primary_segment)
    audience_tokens = _tokens(context.target_audience)
    if audience_tokens and not audience_tokens.intersection(segment_tokens):
        grounding_issues += 1
        issues.append(
            _issue(
                "target_user_drift",
                "user_insight.target_user.primary_segment",
                "The primary user no longer overlaps with the normalized target audience.",
                severity="high",
                blocking=True,
            )
        )
    if context.primary_goal != brief.business_goal:
        grounding_issues += 1
        issues.append(
            _issue(
                "business_goal_drift",
                "context.primary_goal",
                "The normalized business goal differs from the submitted brief.",
                severity="high",
                blocking=True,
            )
        )

    brief_tokens = _tokens(f"{brief.product_description} {context.brief_summary}")
    for index, reason in enumerate(value_proposition.reasons_to_believe):
        if reason.support_status != "brief_supported":
            continue
        capability_tokens = _tokens(reason.capability)
        overlap = capability_tokens.intersection(brief_tokens)
        if not any(ref.startswith("context.") for ref in reason.source_refs) or len(overlap) < 2:
            grounding_issues += 1
            issues.append(
                _issue(
                    "reason_to_believe_not_grounded",
                    f"value_proposition.reasons_to_believe.{index}",
                    "A brief-supported capability must cite context and materially overlap with the submitted product description.",
                    severity="high",
                    blocking=True,
                )
            )

    testability_issues = 0
    for index, priority in enumerate(market_hypothesis.validation_priorities):
        if not (
            MEASURABLE_SIGNAL.search(priority.pass_signal)
            and MEASURABLE_SIGNAL.search(priority.fail_signal)
        ):
            testability_issues += 1
            issues.append(
                _issue(
                    "validation_signal_not_measurable",
                    f"market_hypothesis.validation_priorities.{index}",
                    "Pass and fail signals should include an observable threshold, count, percentage, or time bound.",
                )
            )

    checks = [
        QualityCheck(
            code="structure_contract",
            label="Structure contract",
            status="passed",
            detail="All four strategy objects passed their typed structural contracts.",
        ),
        QualityCheck(
            code="reference_integrity",
            label="Reference integrity",
            status="warning" if reference_issues else "passed",
            detail=f"{reference_issues} invalid or downstream reference(s) found."
            if reference_issues
            else "Every source reference resolves to an allowed upstream field.",
        ),
        QualityCheck(
            code="evidence_inheritance",
            label="Evidence inheritance",
            status="warning" if evidence_issues else "passed",
            detail=f"{evidence_issues} item(s) claim stronger evidence than a cited source."
            if evidence_issues
            else "Downstream evidence does not exceed cited upstream evidence.",
        ),
        QualityCheck(
            code="market_claim_grounding",
            label="Market claim grounding",
            status="warning" if claim_issues else "passed",
            detail=(
                f"{claim_issues} market or comparative claim(s) need review."
                if claim_issues
                else f"{auto_revision_count} inferred claim(s) were reframed as explicit hypotheses before this check."
                if auto_revision_count
                else "No unsupported factual market or comparative claims were detected."
            ),
        ),
        QualityCheck(
            code="decision_continuity",
            label="Decision continuity",
            status="warning" if grounding_issues else "passed",
            detail=f"{grounding_issues} target, value, goal, or product-grounding issue(s) found."
            if grounding_issues
            else "Target user, goal, primary value, and product support remain connected.",
        ),
        QualityCheck(
            code="validation_testability",
            label="Validation testability",
            status="warning" if testability_issues else "passed",
            detail=f"{testability_issues} validation plan(s) need measurable pass and fail signals."
            if testability_issues
            else "Every validation priority has measurable pass and fail signals.",
        ),
    ]

    blocking_count = sum(issue.blocking for issue in issues)
    status = "review_required" if blocking_count else "passed_with_notes" if issues else "passed"
    return StrategyQualityReview(
        status=status,
        issue_count=len(issues),
        blocking_issue_count=blocking_count,
        auto_revision_count=auto_revision_count,
        checks=checks,
        issues=issues,
    )
